src/node_report_utils.cc

```diff
@@ -0,0 +1,306 @@
+      std::string name(buffer.data, size);
+      out << "filename: " << name;
+    }
+  }
+}
+
+/*******************************************************************************
+ * Utility function to walk libuv handles.
+ *******************************************************************************/
+void WalkHandle(uv_handle_t* h, void* arg) {
+  std::string type;
+  std::ostringstream data;
+  write_data* wrap = reinterpret_cast<write_data*>(arg);
+  std::ostream* out = wrap->out;
+  int* indent = wrap->indent;
+  uv_any_handle* handle = reinterpret_cast<uv_any_handle*>(h);
+
+  // List all the types so we get a compile warning if we've missed one,
+  // (using default: supresses the compiler warning).
+  switch (h->type) {
+    case UV_UNKNOWN_HANDLE: type = "unknown"; break;
+    case UV_ASYNC: type = "async"; break;
+    case UV_CHECK: type = "check"; break;
+    case UV_FS_EVENT: {
+      type = "fs_event";
+      reportPath(h, data);
+      break;
+    }
+    case UV_FS_POLL: {
+      type = "fs_poll";
+      reportPath(h, data);
+      break;
+    }
+    case UV_HANDLE: type = "handle"; break;
+    case UV_IDLE: type = "idle"; break;
+    case UV_NAMED_PIPE: type = "pipe"; break;
+    case UV_POLL: type = "poll"; break;
+    case UV_PREPARE: type = "prepare"; break;
+    case UV_PROCESS: {
+      type = "process";
+      data << "pid: " << handle->process.pid;
+      break;
+    }
+    case UV_STREAM: type = "stream"; break;
+    case UV_TCP: {
+      type = "tcp";
+      ReportEndpoints(h, data);
+      break;
+    }
+    case UV_TIMER: {
+      uint64_t due = handle->timer.timeout;
+      uint64_t now = uv_now(handle->timer.loop);
+      type = "timer";
+      data << "repeat: " << uv_timer_get_repeat(&(handle->timer));
+      if (due > now) {
+          data << ", timeout in: " << (due - now) << " ms";
+      } else {
+          data << ", timeout expired: " << (now - due) << " ms ago";
+      }
+      break;
+    }
+    case UV_TTY: {
+      int height, width, rc;
+      type = "tty";
+      rc = uv_tty_get_winsize(&(handle->tty), &width, &height);
+      if (rc == 0) {
+        data << "width: " << width << ", height: " << height;
+      }
+      break;
+    }
+    case UV_UDP: {
+      type = "udp";
+      ReportEndpoints(h, data);
+      break;
+    }
+    case UV_SIGNAL: {
+      // SIGWINCH is used by libuv so always appears.
+      // See http://docs.libuv.org/en/v1.x/signal.html
+      type = "signal";
+      data << "signum: " << handle->signal.signum
+      // node::signo_string() is not exported by Node.js on Windows.
+#ifndef _WIN32
+           << " (" << node::signo_string(handle->signal.signum) << ")"
+#endif
+           << "";
+      break;
+    }
+    case UV_FILE: type = "file"; break;
+    // We shouldn't see "max" type
+    case UV_HANDLE_TYPE_MAX : type = "max"; break;
+  }
+
+  if (h->type == UV_TCP || h->type == UV_UDP
+#ifndef _WIN32
+      || h->type == UV_NAMED_PIPE
+#endif
```

Comment:
Why this `#ifndef`?

Reply:
Because `uv_send_buffer_size()` and `uv_recv_buffer_size()`:
> works for TCP, pipe and UDP handles on Unix and for TCP and UDP handles on Windows.

http://docs.libuv.org/en/v1.x/handle.html